### PR TITLE
Go: Implement connect

### PR DESCRIPTION
**Pull request by aaron-congo**
- adds lib.rs and Go logic to establish standalone and cluster connections
- adds integration test logic to setup/teardown redis servers
- adds integration tests for connect workflows

### CODE REVIEW

File: go/api/base_client.go
Comment (aaron-congo) [inline go/api/base_client.go:40]: Do we want to wrap this error with either a different error type or a message?

File: go/api/base_client.go
Comment (aaron-congo) [inline go/api/base_client.go:68]: Do we want to create a generic GlideError type for errors like this? Or just use the generic error?

Comment (aaron-congo) [reply go/api/base_client.go:68]: Also one negative thing about Close returning an error is that the defer statement is now more complex. If the Close return signature didn't include an error type you could just do `defer client.Close()` after opening the connection. With the error return type, you will get a warning that the error return value is ignored. Instead, you have to do
```go
defer func() {
    err := client.Close()
    if err != nil {
        // ...
    }
}()
```
It would be ideal if we didn't return an error here but maybe that's not an option

File: go/api/base_client.go
Comment (aaron-congo) [reply go/api/base_client.go:40]: Okay looks to me like the Python client does not wrap the error, it just lets it bubble up. So I think this should be fine

File: go/api/base_client.go
Comment (aaron-congo) [reply go/api/base_client.go:68]: when close is called for the other clients they do not check whether they were previously closed, they just terminate all resources regardless. I think I'll remove the error and just return without doing anything if baseClient is nil.

File: go/api/base_client.go
Comment (aaron-congo) [reply go/api/base_client.go:71]: Yeah the other clients check if the client is already closed before attempting to execute a command and throw a ClosingError if they are, so we should do something similar if coreClient is nil

File: go/integTest/error_handling_test.go
Comment (aaron-congo) [reply go/integTest/error_handling_test.go:10]: Yeah the equivalent test to this is basic_client_tries_to_connect_to_wrong_address in Java and it is in ErrorHandlingTest, so I feel like it makes sense to leave this here

Comment (SanHalacogluImproving) [reply go/integTest/error_handling_test.go:10]: Yeah, I don't know I think having an explicit file for error handling does not make sense. Makes more sense to keep to error handling in files where it actually has the context (exceptions for commands should be in command tests etc...) for it but do as you wish minor detail.

Comment (aaron-congo) [reply go/integTest/error_handling_test.go:10]: Yeah I'm fine either way, but I kinda agree so I'll move it to connection_test for now

**Comment by aaron-congo** [comment]
> Seems like this PR is too big to go to upstream why not push lib.rs first get the review on it and then push the go code?
I think I brought up this idea before and IIRC Andrew thought they wouldn't want the lib.rs without any integration tests. The PR isn't small but it isn't massive either, I think I will just open it against upstream once this is approved and see if they ask me to split the lib.rs out

**Review by jonathanl-bq — APPROVED** [review]

File: go/api/base_client.go
Comment (SanHalacogluImproving) [reply go/api/base_client.go:70]: https://github.com/aws/glide-for-redis/pull/1077#discussion_r1514456092 I think it should be a closing error.

**Review by SanHalacogluImproving — APPROVED** [review]

File: .github/workflows/go.yml
Comment (Yury-Fridlyand) [inline .github/workflows/go.yml]: Probably you need to extend it instead of overwritting: `PATH=$PATH:/new/path`

File: go/api/base_client.go
Comment (Yury-Fridlyand) [reply go/api/base_client.go:70]: What if client get closed while a command being executed? We should protect vs application crash in that case too.
My idea so far is to add one more argument to callbacks - client it. A callback submitted to rust would be a static function then, which can dispatch the response to a client if it is not closed yet or ignore vice versa.
TBD

File: go/integTest/setup_teardown_test.go
Comment (Yury-Fridlyand) [inline go/integTest/setup_teardown_test.go:1]: The file name is confusing me - it does not contain any tests.

File: go/integTest/setup_teardown_test.go
Comment (Yury-Fridlyand) [inline go/integTest/setup_teardown_test.go:40]: Sorry for stupid golang question - what does `suite.T()` do? Where it is defined?

File: .github/workflows/go.yml
Comment (Yury-Fridlyand) [inline .github/workflows/go.yml:86]: Upload cluster logs too from `utils/clusters/**`

File: go/integTest/connection_test.go
Comment (Yury-Fridlyand) [inline go/integTest/connection_test.go:23]: Should be able to connect with only 1 port which belongs to any cluster node (even replica)

File: go/src/lib.rs
Comment (Yury-Fridlyand) [inline go/src/lib.rs:1]: Move this after license header

File: .github/workflows/go.yml
Comment (aaron-congo) [reply .github/workflows/go.yml]: Yeah I was wondering about that, currently the path is not set on ubuntu GH runners but it would be better to extend it. Wanted to see if I could get the CI to be green first, but now that it is I'll do this

File: go/integTest/setup_teardown_test.go
Comment (aaron-congo) [reply go/integTest/setup_teardown_test.go:1]: I'm using Testify to add setup/teardown steps to our integration tests (more info [here](https://vidler.app/blog/golang/using-testify-suite-setup-teardown-to-run-operations-before-after-groups-of-tests-in-go/) and [here](https://pkg.go.dev/github.com/stretchr/testify/suite)). The file has one test called TestGlideTestSuite which is actually an entry point to run all the integration tests. To run TestGlideTestSuite the file name needs to end in `_test.go` otherwise Go won't see that TestGlideTestSuite needs to be run. Let me know if you have ideas for a better file name, would `glide_test_suite_test.go` be better?

Comment (Yury-Fridlyand) [reply go/integTest/setup_teardown_test.go:1]: Oh. Thank you for the explanation.
Keep this as is.

File: .github/workflows/go.yml
Comment (Yury-Fridlyand) [reply .github/workflows/go.yml:79]: Please create a task to investiage and fix that

File: go/integTest/setup_teardown_test.go
Comment (aaron-congo) [reply go/integTest/setup_teardown_test.go:40]: For tests that aren't part of a test suite, you need to pass a paramater of type testing.T which provides a test context and methods for printing and logging (more info [here](https://go.dev/doc/tutorial/add-a-test)). Here we are using a test suite, and you access testing.T via suite.T(). This is mentioned in Testify's docs [here](https://pkg.go.dev/github.com/stretchr/testify/suite#Suite.T)

File: go/src/lib.rs
Comment (aaron-congo) [reply go/src/lib.rs:1]: The Rust code exists in this PR to get tests passing but we are reviewing it under a separate PR here: https://github.com/aws/glide-for-redis/pull/1127/files

File: go/integTest/connection_test.go
Comment (aaron-congo) [reply go/integTest/connection_test.go:23]: Sure, I'll add a test for that

File: go/api/config.go
Comment (aaron-congo) [inline go/api/config.go:143]: I realized that the final documentation does not show the unexported fields, so it doesn't make sense to mention baseClientConfiguration here

File: go/api/standalone_client.go
Comment (aaron-congo) [inline go/api/standalone_client.go:15]: In Java this is called CreateClient, but in Go the idiom is that a function called NewRedisClient should be provided if users aren't supposed to initialize a RedisClient struct directly. Also, the Go docs normally will document exported symbols alphabetically but will place New functions immediately underneath the document for the associated struct. We lose this clarity if we call this CreateClient

**Comment by aaron-congo** [comment]
Closing as this is superseded by another PR upstream (#1937)

**Pull request closed without merging**
